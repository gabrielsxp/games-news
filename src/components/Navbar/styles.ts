import styled, { css } from 'styled-components'
import media from 'styled-media-query'
import { rgba } from 'polished'

type NavLinkProps = {
  dark?: boolean
}

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 14rem;
  overflow-x: hidden;
  ${({ theme }) => css`
    z-index: ${theme.layers.overlay};
  `}
  & [class^="Container"] {
    max-height: 100%;
  }
`
export const TopBar = styled.div`
  background: transparent;
  height: 7rem;
  width: 100%;
`
export const TopBarWrapper = styled.div`
  background: inherit;
  display: flex;
  align-items: center;
  flex-direction: row;
  justify-content: flex-start;
  width: 100%;
`

export const BottomBar = styled.div`
  ${({ theme }) => css`
    position: relative;
    background-color: ${theme.colors.primary};
    height: 7rem;
    width: 100vw;
    top: 7rem;
    transition: all 0.1s ease;

    ${media.lessThan('medium')`
      height: 5rem;
      background-color: ${theme.colors.secondary};
    `}
  `}
`
export const NavLink = styled.a<NavLinkProps>`
  ${({ theme, dark }) => css`
    font-size: ${dark ? theme.font.sizes.large : theme.font.sizes.medium};
    color: ${dark ? theme.colors.secondary : theme.colors.white};
    font-weight: bold;
    transition: color 0.3s ease-in-out;
    cursor: pointer;
    &:hover {
      color: ${rgba(dark ? theme.colors.primary : theme.colors.secondary, 0.5)};
    }
  `}
`

export const ContentWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-grow: 1;
    > div > a {
      margin-right: ${theme.spacings.small};
    }
    > div > a:last-of-type {
      margin-right: 0;
    }
  `}
`
export const NavWrapper = styled.nav`
  width: 100%;
  height: 7rem;
  display: flex;
  align-items: center;

  ${media.lessThan('medium')`
      height: 5rem;
    `}
`
export const IconWrapper = styled.button`
  width: 2.4rem;
  height: 2.4rem;
  border: none;
  outline: none;
  background: transparent;
  cursor: pointer;
  margin-left: 0.5rem;
  ${({ theme }) => css`
    > svg {
      width: 100%;
      height: 100%;
      color: ${theme.colors.white};
    }
  `}
`

export const MobileMenu = styled.div`
  width: 100vw;
  height: 100vh;
  overflow-y: auto;
  overflow-x: hidden;
  padding: 7rem 0;
  position: fixed;
  transition: all 0.5s ease-in-out;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  ${({ theme }) => css`
    background-color: ${theme.colors.white};
  `}
  &.display-mobile-menu {
    transform: translateY(-100%);
    opacity: 1;
  }
`

export const MobileMenuWrapper = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  ${({ theme }) => css`
    > a {
      margin-bottom: ${theme.spacings.small};
    }
  `}
`
