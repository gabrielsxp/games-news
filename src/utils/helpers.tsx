import { ThemeProvider } from 'styled-components'
import { render, RenderResult } from '@testing-library/react'
import React, { useState, useEffect } from 'react'
import theme from 'styles/theme'

export const renderWithTheme = (children: React.ReactNode): RenderResult =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>)

type WindowSizeProps = {
  width?: number
  height?: number
}

export function useWindowSize() {
  const [windowSize, setWindowSize] = useState<WindowSizeProps>({
    width: undefined,
    height: undefined
  })

  useEffect(() => {
    function handleResize() {
      setWindowSize({
        width: window?.innerWidth,
        height: window?.innerHeight
      })
    }
    window.addEventListener('resize', handleResize)
    handleResize()
    return () => window.removeEventListener('resize', handleResize)
  }, [])

  return windowSize
}
