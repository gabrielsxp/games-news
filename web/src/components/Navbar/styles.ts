import styled, { css } from 'styled-components'
import media from 'styled-media-query'
import { rgba } from 'polished'
import { NavbarProps } from '.'

type NavLinkProps = {
  dark?: boolean
} & NavbarProps

export const Wrapper = styled.div<NavbarProps>`
  display: flex;
  flex-direction: column;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  overflow: hidden;
  ${({ theme, variant }) => css`
    z-index: ${theme.layers.overlay};
    height: ${variant ? '7rem' : '14rem'};
  `}
  ${media.lessThan('medium')`
      height: 5rem;
  `}
  & [class^="Container"] {
    max-height: 100%;
  }
`
export const TopBar = styled.div<NavbarProps>`
  background: transparent;
  height: 7rem;
  width: 100%;
  ${({ variant }) => css`
    display: ${variant ? 'none' : 'block'};
  `}
`
export const TopBarWrapper = styled.div`
  background: inherit;
  display: flex;
  align-items: center;
  flex-direction: row;
  justify-content: flex-start;
  width: 100%;
`

export const BottomBar = styled.div<NavbarProps>`
  ${({ theme, variant }) => css`
    position: relative;
    background-color: ${variant ? theme.colors.white : theme.colors.primary};
    height: 7rem;
    width: 100vw;
    top: 7rem;
    transition: all 0.1s ease;

    ${media.lessThan('medium')`
      height: 5rem;
      background-color: ${theme.colors.secondary};
    `}
  `}
`
export const NavLink = styled.a<NavLinkProps>`
  ${({ theme, dark, variant }) => css`
    font-size: ${dark ? theme.font.sizes.large : theme.font.sizes.medium};
    color: ${dark || variant ? theme.colors.secondary : theme.colors.white};
    font-weight: bold;
    transition: color 0.3s ease-in-out;
    cursor: pointer;
    &:hover {
      color: ${rgba(dark ? theme.colors.primary : theme.colors.secondary, 0.5)};
    }
  `}
`

export const ContentWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    flex-grow: 1;
    > div > a {
      margin-right: ${theme.spacings.small};
    }
    > div > a:last-of-type {
      margin-right: 0;
    }
    > div:first-child {
      margin-right: ${theme.spacings.small};
    }
  `}
`
export const NavWrapper = styled.nav`
  width: 100%;
  height: 7rem;
  display: flex;
  align-items: center;

  ${media.lessThan('medium')`
      height: 5rem;
    `}
`
export const IconWrapper = styled.button<NavbarProps>`
  width: 2.4rem;
  height: 2.4rem;
  border: none;
  outline: none;
  background: transparent;
  cursor: pointer;
  margin-left: 0.5rem;
  ${({ theme }) => css`
    > svg {
      width: 100%;
      height: 100%;
      color: ${theme.colors.white};
    }
  `}
`

export const MobileMenu = styled.div`
  width: 100vw;
  height: 100vh;
  overflow-y: auto;
  overflow-x: hidden;
  padding: 7rem 0;
  position: fixed;
  transition: all 0.5s ease-in-out;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  transform: translateY(-100%);
  ${({ theme }) => css`
    background-color: ${theme.colors.white};
  `}
  &.display-mobile-menu {
    transform: translateY(0%);
    opacity: 1;
  }
`

export const MobileMenuWrapper = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  ${({ theme }) => css`
    > a {
      margin-bottom: ${theme.spacings.small};
    }
  `}
`
