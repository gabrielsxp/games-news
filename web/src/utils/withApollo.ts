import { withApollo } from 'next-apollo'
import { ApolloClient, InMemoryCache } from '@apollo/client'
import { Post } from 'generated/graphql'

const apolloClient = new ApolloClient({
  uri: 'https://games-news.herokuapp.com/graphql',
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          posts: {
            keyArgs: [],
            merge(existing: Post[] | undefined, incoming: Post[]) {
              return [...(existing ?? []), ...incoming]
            }
          }
        }
      }
    }
  })
})

export default withApollo(apolloClient)
