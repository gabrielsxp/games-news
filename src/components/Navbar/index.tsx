import * as S from './styles'
import Logo from 'components/Logo'
import { Container } from 'components/Container'
import Search from 'components/Search'
import MediaMatch from 'components/MediaMatch'
import { MenuAltRight } from '@styled-icons/boxicons-regular/MenuAltRight'
import { useState } from 'react'

const Navbar = () => {
  const [isMobileMenuOpen, setIsMobileMenuOpen] = useState<boolean>(false)

  return (
    <S.Wrapper>
      <MediaMatch lessThan="medium">
        <S.MobileMenu
          data-testid="menu-mobile-wrapper"
          aria-hidden={isMobileMenuOpen}
          className={isMobileMenuOpen ? 'display-mobile-menu' : ''}
        >
          <S.MobileMenuWrapper>
            <S.NavLink aria-hidden={isMobileMenuOpen} dark>
              Home
            </S.NavLink>
            <S.NavLink aria-hidden={isMobileMenuOpen} dark>
              Latest News
            </S.NavLink>
            <S.NavLink aria-hidden={isMobileMenuOpen} dark>
              Latest Articles
            </S.NavLink>
          </S.MobileMenuWrapper>
        </S.MobileMenu>
      </MediaMatch>
      <MediaMatch greaterThan="medium">
        <S.TopBar>
          <Container>
            <S.TopBarWrapper>
              <Logo />
            </S.TopBarWrapper>
          </Container>
        </S.TopBar>
      </MediaMatch>
      <S.BottomBar>
        <Container>
          <S.NavWrapper>
            <S.ContentWrapper>
              <MediaMatch greaterThan="medium">
                <S.NavLink>Home</S.NavLink>
                <S.NavLink>Latest News</S.NavLink>
                <S.NavLink>Latest Articles</S.NavLink>
              </MediaMatch>
              <MediaMatch lessThan="medium">
                <Logo />
              </MediaMatch>
            </S.ContentWrapper>
            <Search />
            <MediaMatch lessThan="medium">
              <S.IconWrapper
                onClick={() => setIsMobileMenuOpen(!isMobileMenuOpen)}
              >
                <MenuAltRight aria-label="menu icon" />
              </S.IconWrapper>
            </MediaMatch>
          </S.NavWrapper>
        </Container>
      </S.BottomBar>
    </S.Wrapper>
  )
}

export default Navbar
