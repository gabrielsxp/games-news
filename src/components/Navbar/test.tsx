import { screen, fireEvent } from '@testing-library/react'
import { renderWithTheme } from 'utils/helpers'

import Navbar from '.'

describe('<Navbar />', () => {
  it('should render the menu correctly', () => {
    renderWithTheme(<Navbar />)

    expect(screen.getAllByText(/home/i)).toHaveLength(2)
    expect(screen.getAllByText(/browse posts/i)).toHaveLength(2)
    expect(screen.getAllByAltText(/Image of a console joystick/i)).toHaveLength(
      2
    )
    expect(screen.getByLabelText(/menu icon/i)).toBeInTheDocument()
  })
  it('should render the mobile menu when then menu icon is clicked', () => {
    renderWithTheme(<Navbar />)
    expect(
      screen.getByLabelText(/menu icon/i).parentElement
    ).toBeInTheDocument()
    expect(screen.getByTestId('menu-mobile-wrapper')).not.toHaveClass(
      'display-mobile-menu'
    )
    fireEvent.click(screen.getByLabelText(/enu icon/i))

    expect(screen.getByTestId('menu-mobile-wrapper')).toHaveClass(
      'display-mobile-menu'
    )
  })
})
