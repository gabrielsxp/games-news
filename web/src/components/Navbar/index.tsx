import * as S from './styles'
import Logo from 'components/Logo'
import { Container } from 'components/Container'
import Search from 'components/Search'
import MediaMatch from 'components/MediaMatch'
import { MenuAltRight } from '@styled-icons/boxicons-regular/MenuAltRight'
import { useState, useEffect } from 'react'
import useScrollPosition from '@react-hook/window-scroll'
import { useWindowSize } from 'utils/helpers'
import Link from 'next/link'

export type NavbarProps = {
  variant?: boolean
}
type NavbarPositionProps = 'fixed' | 'absolute'

const Navbar = ({ variant = false }: NavbarProps) => {
  const [isMobileMenuOpen, setIsMobileMenuOpen] = useState<boolean>(false)
  const [navbarPosition, setNavbarPosition] = useState(
    'absolute' as NavbarPositionProps
  )
  const [navbarBorder, setNavbarBorder] = useState('none')
  const [navbarTop, setNavbarTop] = useState<number>(0)
  const scrollY = useScrollPosition(30)
  const { width } = useWindowSize()

  useEffect(() => {
    if (typeof width === 'number') {
      if (scrollY > 70 && width > 599) {
        const position = 'fixed' as NavbarPositionProps
        setNavbarPosition(position)
        setNavbarTop(0)
        setNavbarBorder('1px solid rgba(244, 5, 49, 0.3)')
      } else if (scrollY > 70 && width <= 599) {
        const position = 'fixed' as NavbarPositionProps
        setNavbarPosition(position)
        setNavbarTop(0)
        setNavbarBorder('1px solid rgba(0, 0, 0, 0.3)')
      } else if (scrollY <= 70 && width > 599) {
        const position = 'absolute' as NavbarPositionProps
        setNavbarPosition(position)
        setNavbarTop(variant ? 0 : 70)
        setNavbarBorder('1px solid rgba(0, 0, 0, 0.3)')
      } else {
        const position = 'fixed' as NavbarPositionProps
        setNavbarPosition(position)
        setNavbarTop(0)
        setNavbarBorder('1px solid rgba(0, 0, 0, 0.3)')
      }
    } else {
      const position = 'absolute' as NavbarPositionProps
      setNavbarPosition(position)
    }
  }, [scrollY, width, variant])

  return (
    <S.Wrapper variant={variant}>
      <MediaMatch lessThan="medium">
        <S.MobileMenu
          data-testid="menu-mobile-wrapper"
          aria-hidden={isMobileMenuOpen}
          className={isMobileMenuOpen ? 'display-mobile-menu' : ''}
        >
          <S.MobileMenuWrapper>
            <Link href="/" passHref>
              <S.NavLink aria-hidden={isMobileMenuOpen} dark>
                Home
              </S.NavLink>
            </Link>
            <Link href="/posts" passHref>
              <S.NavLink aria-hidden={isMobileMenuOpen} dark>
                Browse Posts
              </S.NavLink>
            </Link>
          </S.MobileMenuWrapper>
        </S.MobileMenu>
      </MediaMatch>
      {!variant && (
        <MediaMatch greaterThan="medium">
          <S.TopBar variant={variant}>
            <Container>
              <S.TopBarWrapper>
                <Logo />
              </S.TopBarWrapper>
            </Container>
          </S.TopBar>
        </MediaMatch>
      )}
      <S.BottomBar
        style={{
          position: navbarPosition,
          top: navbarTop,
          borderBottom: navbarBorder
        }}
        variant={variant}
      >
        <Container>
          <S.NavWrapper>
            <S.ContentWrapper>
              <MediaMatch greaterThan="medium">
                {variant && <Logo />}
              </MediaMatch>
              <MediaMatch greaterThan="medium">
                <Link href="/" passHref>
                  <S.NavLink variant={variant}>Home</S.NavLink>
                </Link>
                <Link href="/posts" passHref>
                  <S.NavLink variant={variant}>Browse Posts</S.NavLink>
                </Link>
              </MediaMatch>
              <MediaMatch lessThan="medium">
                <Logo />
              </MediaMatch>
            </S.ContentWrapper>
            <Search variant={variant} />
            <MediaMatch lessThan="medium">
              <S.IconWrapper
                onClick={() => setIsMobileMenuOpen(!isMobileMenuOpen)}
                variant={variant}
              >
                <MenuAltRight aria-label="menu icon" />
              </S.IconWrapper>
            </MediaMatch>
          </S.NavWrapper>
        </Container>
      </S.BottomBar>
    </S.Wrapper>
  )
}

export default Navbar
