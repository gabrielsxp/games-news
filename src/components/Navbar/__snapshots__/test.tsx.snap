// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> should render the heading 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  background: transparent;
  height: 0.7rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="c0"
>
  <div
    class="c1"
  />
  <h1>
    Navbar
  </h1>
</div>
`;
